cmd/gcrane/cmd/gc.go

```diff
@@ -29,28 +31,57 @@ func init() { Root.AddCommand(NewCmdGc()) }
 // NewCmdGc creates a new cobra.Command for the gc subcommand.
 func NewCmdGc() *cobra.Command {
 	recursive := false
+	before := time.Duration(0)
+	pattern := ""
+
 	cmd := &cobra.Command{
 		Use:   "gc",
 		Short: "List images that are not tagged",
 		Args:  cobra.ExactArgs(1),
 		Run: func(_ *cobra.Command, args []string) {
-			gc(args[0], recursive)
+			gc(args[0], recursive, time.Now().Add(-before), pattern)
 		},
 	}
 
 	cmd.Flags().BoolVarP(&recursive, "recursive", "r", false, "Whether to recurse through repos")
+	cmd.Flags().DurationVarP(&before, "before", "b", time.Duration(0), "Age of the oldest uploaded image to print (ns, us, ms, s, m, h)")
+	cmd.Flags().StringVarP(&pattern, "pattern", "p", "", "Will also collect images with no tags matching this pattern")
```

Comment:
I don't find this behavior 100% intuitive. Given the default behavior of only untagged things being collected, this does make some sense, but I have to pause to think about it. I'd like to think about what this flag behavior should be if we made the untagged nature of `gc` a flag (`--untagged` or something, defaulting to true for backwards compat).

One thing that would be useful here is to print both the digest and tag references for this case. Usually for `gcrane gc`, I'm piping the output into `crane delete`. If we don't delete the tags that match this regex, we won't actually be able to delete the image, since it's still tagged. (This isn't required by the API spec but is a quirk of GCR.)

For the collector, let's also:

```go
for _, tag := range manifest.Tags {
  fmt.Printf("%s:%s\n", repo, tag)
}
```

Reply:
Nice catch for the tag issue. Regarding the behaviour of `--pattern` if we add an `--untagged` you're right, it's unclear. My guess is that it should collect only images tagged with only tags name that don't match the pattern. Reversing it (collecting images that match the pattern) makes it a bit more intuitive, but less convenient. Most of the time you know the pattern that you want to keep, not the one that you want to garbage collect.

Reply:
Can you accomplish what you're trying to do by just deleting the tags you want to ignore? Or do you always want to keep these ignored tags if there is another tag pointing to that image?

@ImJasonH mentioned that the regex thing is not too hard to do with `gcrane ls` and `grep` or `grep -v`.

I would suggest we go ahead and merge the `--grace` part and open an issue to keep this discussion going.

Reply:
OK works for me. I suppose I can do a first pass with `gcrane ls | grep` and then use `grcrane gc` to delete untagged images. Let me refactor everything to get rid of the pattern option.

Reply:
Sounds good, thanks! If for some reason that doesn't work well, we can come back and have a longer discussion on naming and behavior and whatnot, but I didn't want to block the `--grace` flag on that conversation because it seems obviously good and useful 👍